packages/backend-function/src/resolve_secret_banner.ts

```diff
@@ -1,15 +1,15 @@
 /* The code in this line replaces placeholder text in environment variables for secrets with values fetched from SSM, this is a noop if there are no secrets */
 import { SSM } from '@aws-sdk/client-ssm';
-import type { AmplifySecretPaths } from './function_env_translator.js';
+import type { SsmEnvVars } from './function_env_translator.js';
 
 /**
  * The body of this function will be used to resolve secrets for Lambda functions
  */
 export const internalAmplifyFunctionBannerResolveSecrets = async (
   client = new SSM()
 ) => {
-  const envPathObject: AmplifySecretPaths = JSON.parse(
-    process.env.AMPLIFY_SECRET_PATHS ?? '{}'
+  const envPathObject: SsmEnvVars = JSON.parse(
```

Comment:
To confirm:
this secret resolution happens once per worker instance (at startup) not per call right?

if that's the case - can this ever yield stale read?
Couple of scenarios:
1. Deployment has both SSM and function changes. This will presumably reboot lambda workers - do we handle race condition, i.e. SSM is written to before lambda re-boots?
2. What happens if deployment has SSM changes (values) but no lambda changes. We should force lambda redeployment in such case to read new values - do we?

Reply:
3. How is this going to behave if CFN rollback happens?

Reply:
> because of the resource dependency order, the lambda will always be updated before the SSM parameter is created

does this mean lambda that's serving frequent live traffic is always going to see stale value on warm workers?

In general I'm afraid we'll have to introduce SNS topic at some point (or some other push notification mechanism). Eventual consistency might be ok for sandbox/dev envs, but for prod serving live traffic the eager refresh might be a must have.

Reply:
regarding that link above. It seems that some eventual consistency is a must.
Perhaps the TTL would be the cheapest solution that gives some predictable outcomes to customer.
Next one to add to the mix would be notification (if we see demand).
At bare minimum we should document steps/provide a script to "bounce" lambda workers (hacks like here https://stackoverflow.com/questions/50866472/restarting-aws-lambda-function-to-clear-cache )

Reply:
The stabilization period makes sense if lambda is always "bounced" during deployment.

This won't work for scenario nr 2 above I think. I.e. if change set ends up just swapping value in SSM but lambda is skipped due to no changes detected.

Reply:
Yes that's true. In that case, I propose we do a refresh interval of 1 minute and tune / tweak as needed from there. Sounds good?

Reply:
Sounds good.

Reply:
Changes from this discussion were merged in https://github.com/aws-amplify/amplify-backend/pull/982